### PR TITLE
🔐 Add Secure API Extensions

### LINKED ISSUE #19802: Add secure context for server-extensions
```[tasklist]
### Tasks
- [ ] https://github.com/directus/directus/issues/19853
- [ ] https://github.com/directus/directus/issues/19854
- [ ] https://github.com/directus/directus/issues/19855
- [ ] https://github.com/directus/directus/issues/19856
- [ ] https://github.com/directus/directus/issues/19804
- [ ] https://github.com/directus/directus/issues/19921
- [ ] https://github.com/directus/directus/issues/19967
- [ ] https://github.com/directus/directus/issues/19988
```

### PR DESCRIPTION
**Pull request by Nitwel**
Due to the release of the marketplace, there is the need for secure api extensions.
This PR allows to configure an extension to being run securely inside an isolated environment.

Features:
- Run extensions securely inside the api
- Configure Permissions to what extensions will have access to
- Be able to access recources like database or make web requests
- Timeout extensions if they run for too long

This PR should be merged before: #19506

Closes https://github.com/directus/directus/issues/19802

### CODE REVIEW

File: api/package.json [hunk 1]
```diff
@@ -156,6 +157,7 @@
 		"snappy": "7.2.2",
 		"stream-json": "1.7.5",
 		"strip-bom-stream": "5.0.0",
+		"tar": "^6.1.13",
```
Comment (DanielBiegler) [inline api/package.json:160]: I assume we probably want to pin the version like the rest?

**Comment by DanielBiegler** [comment]
Note to reader: Because we throw exceptions in the isolate we cant present it to the app like we used to:

image

Context: This is a filter hook denying deletion with:

```js
        const DeleteError = createError('DELETE_PROJECT_ERROR', 'fail delete', 400);
        throw new DeleteError();
```

**Comment by DanielBiegler** [comment]
TODO: Crashing in my hook crashes my whole Api and forces me to restart:

Hook:

```ts
    filter('project.items.delete', () => {
        const DeleteError = createError('DELETE_PROJECT_ERROR', 'fail delete', 400);
        throw new DeleteError();
    });
```

App:

```
Error: socket hang up
    at connResetException (node:internal/errors:717:14)
    at Socket.socketOnEnd (node:_http_client:526:23)
    at Socket.emit (node:events:525:35)
    at endReadableNT (node:internal/streams/readable:1359:12)
    at process.processTicksAndRejections (node:internal/process/task_queues:82:21)
```

Api:

```
node:internal/process/promises:288
            triggerUncaughtException(err, true /* fromPromise */);
            ^

Error: fail delete
    at <redacted>/api/extensions/directus-extension-dan-testing/dist/api.js:1:469
    at <isolated-vm>:11:6
    at (<isolated-vm boundary>)
    at handler (file://<redacted>/api/src/extensions/vm-functions/defineHook/node.ts:1:860)
    at emitFilter (file://<redacted>/api/src/emitter.ts:1:969)
    at deleteMany (file://<redacted>/api/src/services/items.ts:1:19085)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)

Node.js v18.16.1
```

File: api/src/extensions/vm.ts [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+		this.vmFunctions.push(new ConsoleVMFunction());
+		this.vmFunctions.push(new LoggerVMFunction());
+		this.defineEndpoint = new DefineEndpointVMFunction(this.extensionManager);
+		this.defineHook = new DefineHookVMFunction(this.extensionManager);
+		this.defineOperation = new DefineOperationVMFunction();
+	}
+
+	async runExtension(extension: ApiExtensionInfo, extensionPath: string) {
+		const isolateSizeMb = 8;
+		const scriptTimeoutMs = 1000;
```
Comment (DanielBiegler) [inline api/src/extensions/vm.ts:47]: Lets not forget to provide customizability after we're done testing

**Comment by DanielBiegler** [comment]
TODO: my bundle now doesnt load but says it does

```
[12:23:56.067] WARN: Couldn't register bundle "directus-extension-dan-testing"
[12:23:56.067] WARN: Unexpected identifier [extensionLoader.js:1:98]
    err: {
      "type": "SyntaxError",
      "message": "Unexpected identifier [extensionLoader.js:1:98]",
      "stack":
          SyntaxError: Unexpected identifier [extensionLoader.js:1:98]
              at (<isolated-vm boundary>)
              at run (file:///home/dan/Projects/directus-src/api/src/extensions/vm.ts:1:3702)
    }
[12:23:56.068] INFO: Enabled extensions: directus-extension-dan-testing
[12:23:56.068] INFO: Disabled extensions: 
[12:23:56.116] INFO: Server started at http://0.0.0.0:8055
```

Bundle doesnt load and the hook inside it doesnt fire.

**Comment by DanielBiegler** [comment]
Would be check off #19612 with this PR

File: api/src/app.ts [hunk 3]
```diff
@@ -272,6 +273,7 @@ export default async function createApp(): Promise<express.Application> {
 	app.use('/collections', collectionsRouter);
 	app.use('/dashboards', dashboardsRouter);
 	app.use('/extensions', extensionsRouter);
+	app.use('/extension_permissions', extensionPermissionsRouter);
```
Comment (rijkvanzanten) [inline api/src/app.ts:276]: Lets make this `/extensions/permissions`

File: api/src/controllers/extension-permissions.ts [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+import { MetaService } from '../services/meta.js';
+import type { PrimaryKey } from '../types/index.js';
+import asyncHandler from '../utils/async-handler.js';
+import { sanitizeQuery } from '../utils/sanitize-query.js';
+import useCollection from '../middleware/use-collection.js';
+import { getExtensionManager } from '../extensions/extensions.js';
+
+const router = express.Router();
+
+router.use(useCollection('directus_extension_permissions'));
```
Comment (rijkvanzanten) [inline api/src/controllers/extension-permissions.ts:16]: Any chance we can combine this with `directus_extensions`?

Comment (rijkvanzanten) [reply api/src/controllers/extension-permissions.ts:16]: I don't know if this is possible, but it'd be nice if `permissions` was a column of `directus_extensions`

File: api/src/controllers/extension-permissions.ts [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+
+		const extensionManager = await getExtensionManager();
+		await extensionManager.reloadExtensionPermissions();
+
+		return next();
+	}),
+	respond
+);
+
+const readHandler = asyncHandler(async (req, res, next) => {
```
Comment (rijkvanzanten) [inline api/src/controllers/extension-permissions.ts:62]: Unrelated to this PR, but man we should consolidate all this boilerplate into a single "create standard controller"

File: api/src/controllers/extensions.ts [hunk 6]
```diff
@@ -1,29 +1,25 @@
...
 
-		if (!isIn(type, EXTENSION_TYPES)) {
-			throw new RouteNotFoundError({ path: req.path });
-		}
-
-		const extensionManager = getExtensionManager();
+		const extensionManager = await getExtensionManager();
 
-		const extensions = extensionManager.getExtensionsList(type);
+		const extensions = extensionManager.getDisplayExtensions();
```
Comment (rijkvanzanten) [inline api/src/controllers/extensions.ts:22]: `/` route is now for just Display extensions?

File: api/src/errors/extension-permissions.ts [hunk 7]
```diff
@@ -0,0 +1,16 @@
+import { createError } from '@directus/errors';
+import { ErrorCode } from './codes.js';
+
+interface ExtensionServiceErrorExtensions {
+	service: string;
+	reason: string;
+}
+
+export const messageConstructor = ({ service, reason }: ExtensionServiceErrorExtensions) =>
+	`No access to "${service}". ${reason}.`;
```
Comment (rijkvanzanten) [inline api/src/errors/extension-permissions.ts:10]: Lets add the operation here as well.

File: api/src/extensions/installation.ts [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+			throw new Error(`The name "${name}" is reserved for internal use.`);
+		}
+
+		const extension = this.extensionManager.getExtension(name);
+
+		if (extension) {
+			throw new Error(`Extension "${name}" is already installed.`);
+		}
+
+		const axios = (await import('axios')).default;
```
Comment (rijkvanzanten) [inline api/src/extensions/installation.ts:29]: Async import is no longer required now that the API codebase is ESM

File: api/src/controllers/extension-permissions.ts (hunk 4 above)
Comment (paescuj) [reply api/src/controllers/extension-permissions.ts:16]: Would be nice!
Also if we aren't able to combine it, IMO `directus_extensions_permission` would look a tad cleaner. (I know very nitpicking 🤓)

Comment (paescuj) [reply api/src/controllers/extension-permissions.ts]: Resolved in https://github.com/directus/directus/pull/19507/commits/f7e361e238300b1018e56f72a2de0529c2babef6

**Comment by rijkvanzanten** [comment]
This is superseded by #20031

**Pull request closed without merging**